### PR TITLE
Make sslConfig agnostic of ssl-enabled DCs and enable router ssl tests

### LINKED ISSUE #447: Fix router - server SSL integration test
`RouterServerSSLTest` isn't doing any tests related to SSL due to two reasons
1. The configuration is set up incorrectly. The `NonBlockingRouter` does not respect the `ssl.enabled.datacenters` value but instead relies on the `ClusterMap` to give it the correct port based on `clustermap.ssl.enabled.datacenters`. Fixing this will require changes to `com.github.ambry.server.MockClusterMap` and requires some refactoring.
2. The router needs to ensure that requests *surely* go through the ssl pathway and it should not be a *might*.

### PR DESCRIPTION
**Pull request by pnarayanan**
- Remove ssl-enabled DC parameter from sslConfig and rely on the one in
the ClusterMapConfig (which is the right place for it).
- Reenable Router-Server SSL tests.

Fixes #447 and #448

### CODE REVIEW

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java [hunk 1]
```diff
@@ -56,6 +51,32 @@ private void populatePorts(ArrayList<Port> ports) {
     }
   }
 
+  /**
+   * Set the datacenters to which ssl is enabled. If the datacenter on which this datanode resides is
+   * part of the datacenters to which ssl is enabled, then this datanode will always return an SSL port in
```
Comment (vgkholla) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java]: is there a way (or necessity) to ensure the other way round? i.e. if a node is in a DC that is ssl enabled for everyone else, then it should get an SSL port for any node that it tries to connect to.

Putting it another way, is there a way to customize this clustermap per DC? The old `getPortToConnectTo(List<String>)` provided that flexibility.

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java [hunk 2]
```diff
@@ -36,20 +41,41 @@
...
+  public void before() {
+    for (MockDataNodeId m : clusterMap.getDataNodes()) {
+      m.setSslPortRequestedStatus(false);
+    }
+  }
+
+  @After
+  public void after() {
+    for (MockDataNodeId m : clusterMap.getDataNodes()) {
+      if (m.sslPortWasRequested()) {
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java]: This seems like a weak check that might not hold if we fix some things I think.

To elaborate, it seems like the `Router` is the only one using `getPortToConnectTo()` currently. But this need not always stay true. For e.g. [this](https://github.com/linkedin/ambry/blob/master/ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java#L374) looks like duplicate code to me and this function can easily be discarded and  the single call to this function can be replaced by `getPortToConnectTo()`. (@nsivabalan - could you comment on the feasibility of removing the linked function? It looks superfluous to me).

I'm not saying that the change to `ReplicationManager` has to go into this patch (it would be great if it could!), but in the case of this test, since it also brings up servers, making that replacement would make this check invalid. What do you think?

File: ambry-network/src/test/java/com.github.ambry.network/BlockingChannelConnectionPoolTest.java [hunk 3]
```diff
@@ -377,7 +387,7 @@ public void testBlockingChannelConnectionPool()
     props.put("connectionpool.max.connections.per.port.ssl", "5");
     ConnectionPool connectionPool =
         new BlockingChannelConnectionPool(new ConnectionPoolConfig(new VerifiableProperties(props)), sslConfig,
-            new MetricRegistry());
+            plainTextClusterMapConfig, new MetricRegistry());
```
Comment (pnarayanan) [inline ambry-network/src/test/java/com.github.ambry.network/BlockingChannelConnectionPoolTest.java]: the other test is in disabled state due to errors - just made the change anyway to ensure that we do the right thing when we get around to fixing that.

File: ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java [hunk 4]
```diff
@@ -51,27 +51,31 @@
...
-          String sslEnabledDatacenters = getSSLEnabledDatacenterValue(dataNodeId.getDatacenterName(), datacenterList);
-          sslProps.setProperty("ssl.enabled.datacenters", sslEnabledDatacenters);
+        if (sslEnabledDataCentersStr != null) {
+          dataNodeId.setSslEnabledDataCenters(sslEnabledDataCenterList);
         }
         initializeServer(dataNodeId, sslProps, enableHardDeletes, time);
       }
-    } catch (InstantiationException e) {
+    } catch (InstantiationException e)
+
```
Comment (pnarayanan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java]: fixed.

File: config/frontend.properties [hunk 5]
```diff
@@ -15,6 +15,6 @@ rest.server.blob.storage.service.factory=com.github.ambry.frontend.AmbryBlobStor
 
 # router
 router.hostname=localhost
-router.datacenter.name=Datacenter
```
Comment (pnarayanan) [inline config/frontend.properties]: fixed.

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java (hunk 1 above)
Comment (pnarayanan) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java]: customizing is hard when the same clustermap is used by all nodes. The clustermap gets initialized with a config and that config has to be different for the clustermap of each node. For the main code, this makes things cleaner. In tests it would be cumbersome to create too many clustermaps unfortunately.

Comment (pnarayanan) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java]: the old way provided the flexibility only in tests.

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java (hunk 2 above)
Comment (pnarayanan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java]: Made the change in the `ReplicationManager` as well.

This check is weak but it is really to ensure that SSL tests are indeed doing what they intend to do, as that wasn't happening earlier and it went undetected.

The check for whether SSL is being exercised when asked to do SSL is not really being done in the integration tests. I think it would be hard to do in the integration tests, but unit tests should over that.

**Comment by pnarayanan** [comment]
About your other comments - `ToolsUtil.createSSLProperties()` is already taken care of.
Got rid of the deprecated method - that was always the plan, but required changes in tools and such.

**Comment by pnarayanan** [comment]
@vgkholla Updated. Built and tested on OS X and Linux.

File: config/frontend.ssl.properties [hunk 6]
```diff
@@ -17,7 +17,6 @@ rest.server.blob.storage.service.factory=com.github.ambry.frontend.AmbryBlobStor
 # router
 router.hostname=localhost
 router.datacenter.name=Datacenter
```
Comment (cgtz) [inline config/frontend.ssl.properties]: This should stay as "Datacenter", right? Since the other one was changed from "Data-Center" to "Datacenter."

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java [hunk 7]
```diff
@@ -56,6 +51,32 @@ private void populatePorts(ArrayList<Port> ports) {
...
+   */
+  public void setSslPortRequestedStatus(boolean status) {
+    sslPortRequested = status;
+  }
+
+  /**
+   * Get the status for whether an SSL port was requested from this datanode.
+   * @return true if SSL port was requested; false otherwise.
+   */
+  public boolean sslPortWasRequested() {
```
Comment (nsivabalan) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java]: why is this required ?

File: ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java [hunk 8]
```diff
@@ -288,7 +287,7 @@ public ReplicationManager(ReplicationConfig replicationConfig, SSLConfig sslConf
       this.notification = requestNotification;
       this.metricRegistry = metricRegistry;
       this.dataNodeRemoteReplicaInfosPerDC = new HashMap<String, DataNodeRemoteReplicaInfos>();
-      this.sslEnabledDatacenters = Utils.splitString(sslConfig.sslEnabledDatacenters, ",");
+      this.sslEnabledDatacenters = Utils.splitString(clusterMapConfig.clusterMapSslEnabledDatacenters, ",");
```
Comment (nsivabalan) [inline ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java]: those will be taken as DCname (which is considered to have a whitespace). So, may not work as expected

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java (hunk 1 above)
Comment (nsivabalan) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java]: @vgkholla : Can you elaborate a bit on your comment ? Do you mean to suggest having different clustermaps for different DCs. For instance, for DC1, the ssl list is (DC2, DC3), for DC2, the ssl list is (DC1,DC3) and for DC3 ssl list is (DC1, DC2). Are you asking why dont we do this in our tests and we just have one clustermap for all which has (DC1, D2, DC3) ?

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java (hunk 2 above)
Comment (nsivabalan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java]: I guess we need something of this sort to confirm that ssl was the way the communication happened. But I am concerned about having it in after clause. I mean, how can we ensure that all of the tests will involve every data node. If in future we add tests which doesn't poll every data node this check would fail right?

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java [hunk 9]
```diff
@@ -27,8 +27,8 @@
   List<String> mountPaths;
   String hostname = "localhost";
   String datacenter;
-  long rackId = -1;
   ArrayList<String> sslEnabledDataCenters = new ArrayList<String>();
+  boolean sslPortRequested = false;
```
Comment (pnarayanan) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java]: fixed

File: ambry-network/src/test/java/com.github.ambry.network/TestSSLUtils.java [hunk 10]
```diff
@@ -202,7 +201,7 @@ public static void addSSLProperties(Properties props, String sslEnabledDatacente
     props.put("ssl.truststore.path", trustStoreFile.getPath());
     props.put("ssl.truststore.password", TRUSTSTORE_PASSWORD);
     props.put("ssl.cipher.suites", SSL_CIPHER_SUITES);
-    props.put("ssl.enabled.datacenters", sslEnabledDatacenters);
+    props.put("clustermap.ssl.enabled.datacenters", sslEnabledDatacenters);
   }
 
   /**
```
Comment (pnarayanan) [inline ambry-network/src/test/java/com.github.ambry.network/TestSSLUtils.java]: fixed

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java (hunk 2 above)
Comment (pnarayanan) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerSSLTest.java]: @nsivabalan @vgkholla's comment is that this check does not necessarily ensure that the router communicated over SSL to the datanodes - because the ssl ports could be requested by the servers as part of replication - and in these tests they all share the clustermap.

**Comment by pnarayanan** [comment]
@vgkholla @cgtz @nsivabalan addressed comments, and made changes to authoritatively determine if router communicates over SSL or not. Please take a look at the latest commit.

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java (hunk 1 above)
Comment (vgkholla) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockDataNodeId.java]: @nsivabalan yup - or some way for the clustermap to figure out which DC the user of the clustermap is in. This is pretty hard to do from what I can see of the code (and from @pnarayanan's comment above)  - but would have been a nice to have.

**Comment by vgkholla** [comment]
LGTM. Please verify build and formatting. Thanks for making this change ! Avoids a lot of confusion.

**Comment by pnarayanan** [comment]
Had made a tiny change after testing out everything, which had a bug that fails the test. Just pushed the fix.
Formatted; built and tested on Linux and OS X.

**Comment by cgtz** [comment]
LGTM

I can merge after @nsivabalan takes a look
